bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/FindReplaceLogic.java

```diff
@@ -0,0 +1,799 @@
+			findReplacePosition += r.y;
+
+		nextReplactionOperationNeedsFindOperationFirst = false;
+
+		int index = findIndex(findString, findReplacePosition);
+
+		if (index == -1) {
+			String msg = NLSUtility.format(EditorMessages.FindReplace_Status_noMatchWithValue_label, findString);
+			statusMessage(false, EditorMessages.FindReplace_Status_noMatch_label, msg);
+			return false;
+		}
+
+		if (forwardSearch && index >= findReplacePosition || !forwardSearch && index <= findReplacePosition)
+			statusMessage(""); //$NON-NLS-1$
+
+		return true;
+	}
+
+	public boolean performFindFirstThenReplaceInASecondStep(String findString, String replaceString) {
+		if (nextReplactionOperationNeedsFindOperationFirst) {
+			performSearch(findString);
+		}
+		if (performReplaceSelection(replaceString)) {
+			performSearch(findString);
+			return true;
+		}
+		return false;
+	}
+
+	public boolean performSelectAndReplace(String findString, String replaceString) {
+		if (nextReplactionOperationNeedsFindOperationFirst)
+			performSearch(findString);
+		return performReplaceSelection(replaceString);
+	}
+
+	public void updateTarget(IFindReplaceTarget newRarget, boolean canEditTarget) {
+		this.isTargetEditable = canEditTarget;
+		nextReplactionOperationNeedsFindOperationFirst = true;
+
+		if (this.target != newRarget) {
+			if (newRarget != null && newRarget instanceof IFindReplaceTargetExtension)
+				((IFindReplaceTargetExtension) newRarget).endSession();
+
+			this.target = newRarget;
+			if (newRarget != null)
+				fIsTargetSupportingRegEx = newRarget instanceof IFindReplaceTargetExtension3;
+
+			if (newRarget instanceof IFindReplaceTargetExtension) {
+				((IFindReplaceTargetExtension) newRarget).beginSession();
+
+				setGlobalSearch(true);
+			}
+		}
+
+		initIncrementalBaseLocation();
+	}
+
+	public void endSession() {
+		if (target != null && target instanceof IFindReplaceTargetExtension)
+			((IFindReplaceTargetExtension) target).endSession();
+
+		target = null;
+	}
+
+
+	public void deactivateScope() {
+		if (target != null && (target instanceof IFindReplaceTargetExtension))
+			((IFindReplaceTargetExtension) target).setScope(null);
+
+		oldScope = null;
+	}
+
+	public String getCurrentSelection() {
+		if (target == null) {
+			return null;
+		}
+
+		return target.getSelectionText();
+	}
+
+	/**
+	 * Returns whether the target is editable.
+	 *
+	 * @return <code>true</code> if target is editable
+	 */
+	public boolean isEditable() {
+		boolean isEditable = (target == null ? false : target.isEditable());
+		return isTargetEditable && isEditable;
+	}
+
+	/**
+	 * @return Whether the target supports multiple selections
+	 */
+	public boolean supportsMultiSelection() {
+		return target instanceof IFindReplaceTargetExtension4;
+	}
+
+	/**
+	 * @return Whether the target is available
+	 */
+	public boolean isTargetAvailable() {
+		return target != null;
+	}
```

Comment:
These methods should be `private` as they do not provide functionality semantically related to a search.